# Nivel 1-2: Muy básicos (entender y contar lo obvio)

## Ejercicio 1: Notas de un Examen


**Datos:** Lista de notas (0-10) de una clase.

notas = [7, 8, 6, 9, 7, 5, 8, 7, 6, 4, 9, 8, 7, 10, 6, 7, 8, 5, 9, 7]

**Tareas:**
1. Calcular media, mediana y moda
2. Responder por escrito:
   - ¿La nota "típica" del grupo es más cercana a la media o a la mediana?
   - ¿Qué medida usarías para explicar el rendimiento general al claustro? ¿Por qué?

In [3]:
import pandas as pd

minutos = [7, 8, 6, 9, 7, 5, 8, 7, 6, 4, 9, 8, 7, 10, 6, 7, 8, 5, 9, 7]

series = pd.Series(minutos)

print("Media: ", series.mean())
print("Mediana: ", series.median())
print("Moda: ", series.mode()[0])

Media:  7.15
Mediana:  7.0
Moda:  7


En este caso salen muy parecidas porque no hay resultados polarizados, pero en todo caso yo usaria la mediana para evitar los extremos

## Ejercicio 2: Horas de Estudio Semanales

Se realizó una encuesta a un grupo de estudiantes sobre la cantidad de horas que dedican a estudiar por semana. Los datos obtenidos son:

**Datos**: 5, 8, 6, 7, 50, 6, 8, 7, 5, 6

## Tareas:
1. Calcular la **media** y la **mediana**
2. **Comparar** ambas medidas
3. **Explicar** qué dice eso sobre el hábito de estudio del grupo

💡 **Pista didáctica**: Observe cómo los valores extremos pueden "engañar" a la media

In [4]:
minutos = [5, 8, 6, 7, 50, 6, 8, 7, 5, 6]

series = pd.Series(minutos)

print("Media: ", series.mean())
print("Mediana: ", series.median())

Media:  10.8
Mediana:  6.5


La media no se ajusta a la mediana, hay algún valor mucho mayor dado a que media >>> mediana

# Nivel 3–4 · Introducir dispersión y variabilidad

## Ejercicio 3: Sueldos en una empresa pequeña

En una empresa pequeña se registraron los sueldos mensuales de sus empleados. Uno de ellos (el gerente) tiene un sueldo muy alto comparado con el resto.

**Datos (ejemplo)**: 1200, 1400, 1350, 1500, 1600, 1450, 1300, 1250, 1800, 10000

## Tareas
- Calcular **media**, **mediana** y **desviación estándar**.
- Explicar:
  - ¿La media representa bien al "empleado medio"?
  - ¿Qué nos dice la desviación estándar sobre la **desigualdad salarial**?

💡 Pista didáctica: el sueldo del gerente es un **valor extremo** que puede **sesgar la media**.

In [5]:
minutos = [1200, 1400, 1350, 1500, 1600, 1450, 1300, 1250, 1800, 10000]

series = pd.Series(minutos)

print("Media: ", series.mean())
print("Mediana: ", series.median())
print("Desviación estándard: ", series.std())

Media:  2285.0
Mediana:  1425.0
Desviación estándard:  2716.518400862071


La desviación estandard es muy elevanda, y la solo viendo como la media es un 50% mayor a la mediana es aparecente de que hay una brecha salarial sustancial

## Ejercicio 4: Tiempo para llegar al trabajo
Se registraron los **minutos de desplazamiento** diario al trabajo para un grupo de empleados.

**Datos (ejemplo)**: 15, 20, 18, 22, 19, 17, 45, 60, 16, 21

## Tareas
- Calcular **media**, **mediana** y **desviación estándar**.
- Interpretar:
  - ¿El grupo es **homogéneo** o hay **grandes diferencias**?
  - ¿Qué tipo de trabajador tiene **peor experiencia diaria**?

💡 Pista didáctica: algunos trayectos largos (valores extremos) pueden **inflar la media** y aumentar la **desviación estándar**.

In [6]:
minutos = [15, 20, 18, 22, 19, 17, 45, 60, 16, 21]

series = pd.Series(minutos)

print("Media: ", series.mean())
print("Mediana: ", series.median())
print("Desviación estándard: ", series.std())

Media:  25.3
Mediana:  19.5
Desviación estándard:  14.922391824964851


Una desviación de casi 15 minutos es bastante, y viendo como la diferencia entre el minimo y el máximo es de 45 minutos hay trabajadores que están haciendo horas extra solamente yendo al puesto de trabajo

#  Nivel 5–6 · Outliers y análisis crítico

##  Ejercicio 5

Se registraron las **ventas diarias** de una tienda a lo largo de un **mes**. Hubo **un día con ventas muy altas** respecto al resto.

## Tareas
- Detectar **posibles outliers**.
- Calcular **media** y **mediana** **con y sin** ese valor.
- Explicar **qué pudo pasar** ese día (rebajas, campaña, error…).

🎯 Storytelling: "**Un solo día excepcional** distorsiona la **percepción** del mes completo".

In [27]:
import numpy as np
import plotly.express as px

ventas = np.random.uniform(100, 1000, 30)

ventas = np.append(ventas, 2500)

df = pd.DataFrame({"Ventas": ventas})

Q1 = df["Ventas"].quantile(0.25)
Q3 = df["Ventas"].quantile(0.75)
IQR = Q3 - Q1

lim_inf = Q1 - 1.5 * IQR
lim_sup = Q3 + 1.5 * IQR

df["Outlier"] = (df["Ventas"] < lim_inf) | (df["Ventas"] > lim_sup)

print("Outliers detectados:")
print(df[df["Outlier"]])

print("Media con outlier: ", df["Ventas"].mean())
print("Mediana con outlier: ", df["Ventas"].median())

df_limpio = df[(df["Ventas"] >= lim_inf) & (df["Ventas"] <= lim_sup)]

print("Media sin outlier: ", df_limpio["Ventas"].mean())
print("Mediana sin outlier: ", df_limpio["Ventas"].median())

fig = px.box(df, y="Ventas", points="outliers", title="Boxplot de Ventas")
fig.show()

Outliers detectados:
    Ventas  Outlier
30  2500.0     True
Media con outlier:  589.9384087857353
Mediana con outlier:  527.576451384934
Media sin outlier:  526.2696890785932
Mediana sin outlier:  521.6422545677825


## Ejercicio 6: Uso del móvil al día

Se registraron las **horas diarias de uso del móvil** de un grupo de personas durante un periodo reciente.

## Tareas
- Calcular **media**, **mediana** y **desviación estándar**.
- **Identificar outliers** (usos extremadamente altos o bajos).
- Explicar si existe un **uso normal** o varios **perfiles de usuarios**.

🎯 Pista didáctica: valores muy altos de uso pueden sesgar la media y aumentar la variabilidad percibida.

# Nivel 7–8 · Diagramas de caja (bigotes)

## Ejercicio 7: Notas por trimestre


Se tienen las **notas del mismo grupo** en **3 trimestres**.

trimestre1 = [5.5, 6.0, 5.8, 6.2, 7.0, 6.5, 5.9, 6.1, 7.2, 6.8, 5.4, 6.0, 6.3]

trimestre2 = [6.2, 6.5, 6.8, 7.1, 7.3, 6.9, 7.0, 6.7, 7.5, 7.4, 6.6, 7.2, 7.1]

trimestre3 = [7.0, 7.2, 7.5, 7.8, 8.0, 7.9, 7.6, 7.7, 8.2, 8.1, 7.4, 7.3, 7.9]

## Tareas
- Crear un **diagrama de caja** por trimestre.
- Comparar **mediana** y **rango intercuartílico (IQR)** de cada trimestre.
- Explicar la **evolución del grupo**.

🎯 Storytelling: "No solo sube la **nota central**, también se **reduce la dispersión**…"

In [7]:
import plotly.express as px

trimestre1 = [5.5, 6.0, 5.8, 6.2, 7.0, 6.5, 5.9, 6.1, 7.2, 6.8, 5.4, 6.0, 6.3]
trimestre2 = [6.2, 6.5, 6.8, 7.1, 7.3, 6.9, 7.0, 6.7, 7.5, 7.4, 6.6, 7.2, 7.1]
trimestre3 = [7.0, 7.2, 7.5, 7.8, 8.0, 7.9, 7.6, 7.7, 8.2, 8.1, 7.4, 7.3, 7.9]

data = pd.DataFrame({
    "Trimestre 1": trimestre1,
    "Trimestre 2": trimestre2,
    "Trimestre 3": trimestre3
})

# Convertir a formato largo
df_long = data.melt(var_name="Trimestre", value_name="Calificación")

fig = px.box(df_long, x="Trimestre", y="Calificación", title="Boxplot por Trimestre")
fig.show()

for col in data.columns:
    q1 = data[col].quantile(0.25)
    q3 = data[col].quantile(0.75)
    iqr = q3 - q1
    mediana = data[col].median()
    print(f"{col} → Mediana: {mediana}, IQR: {iqr}")

Trimestre 1 → Mediana: 6.1, IQR: 0.5999999999999996
Trimestre 2 → Mediana: 7.0, IQR: 0.5
Trimestre 3 → Mediana: 7.7, IQR: 0.5


Se ve una evolución claramente positiva dado a que la mediana aumenta y el IRQ se reduce, por tanto hay más beneficios de manera más consistente

## Ejercicio 8: Consumo de datos móviles

Se registró el **consumo de datos (GB)** de distintos usuarios durante un periodo reciente.

## Consumo de datos (GB) por usuario; incluye algunos consumidores intensivos
consumo_gb = [4.2, 3.8, 5.1, 2.9, 3.5, 4.0, 6.2, 5.5, 3.1, 2.7,
              12.0, 15.5, 4.3, 3.9, 5.0, 6.8, 7.2, 2.5, 3.0, 4.8]

## Tareas
- Crear un **boxplot** del consumo.
- Identificar **outliers**.
- Clasificar **perfiles**: bajo, medio, intensivo.

💡 Aquí se ve muy bien el valor visual del **bigote** para detectar extremos sin usar gráficos complejos.

In [8]:
consumo_gb = [4.2, 3.8, 5.1, 2.9, 3.5, 4.0, 6.2, 5.5, 3.1, 2.7,
              12.0, 15.5, 4.3, 3.9, 5.0, 6.8, 7.2, 2.5, 3.0, 4.8]

df = pd.DataFrame({"Consumo": consumo_gb})


fig = px.box(df, y="Consumo", points="outliers", title="Boxplot Consumo de Datos Móviles")
fig.show()

print(f"Perfil bajo hasta {df["Consumo"].quantile(0.25)} Gb")
print(f"Perfil medio hasta {df["Consumo"].quantile(0.5)} Gb")
print(f"Perfil alto a partir de {df["Consumo"].quantile(0.75)} Gb")

Perfil bajo hasta 3.4 Gb
Perfil medio hasta 4.25 Gb
Perfil alto a partir de 5.675 Gb


# Nivel 9–10 · Storytelling completo (tipo mini-informe)

## Ejercicio 9: Comparación entre dos grupos (mañana vs tarde)

Se registraron las **horas diarias de estudio** de dos grupos del mismo curso:
- **Grupo A (mañana)**
- **Grupo B (tarde)**

## Datos sintéticos: medias similares, variabilidad distinta y algunos extremos en B
A = [6.5, 7.0, 6.8, 7.2, 6.9, 7.1, 6.7, 7.0, 6.6, 6.9, 7.3, 6.8, 7.0, 6.7, 7.1]

B = [6.4, 7.1, 3.0, 9.5, 6.8, 7.0, 2.5, 8.8, 6.9, 7.2, 10.0, 5.5, 6.6, 7.4, 4.2]

## Tareas
- Calcular **todas las medidas** descriptivas principales.
- Crear **diagramas de caja** comparativos.
- Escribir una **conclusión**: ¿qué grupo es más **constante**? ¿cuál tiene más **casos extremos**?

🎯 Storytelling: "Aunque ambos grupos estudian una **media similar**, su **comportamiento** es muy distinto…"

In [10]:
A = [6.5, 7.0, 6.8, 7.2, 6.9, 7.1, 6.7, 7.0, 6.6, 6.9, 7.3, 6.8, 7.0, 6.7, 7.1]

B = [6.4, 7.1, 3.0, 9.5, 6.8, 7.0, 2.5, 8.8, 6.9, 7.2, 10.0, 5.5, 6.6, 7.4, 4.2]

print(f"Media del grupo A: {sum(A)/len(A)}")
print(f"Media del grupo B: {sum(B)/len(B)}")

dfA = pd.DataFrame({"Notas": A})
dfB = pd.DataFrame({"Notas": B})

fig = px.box(dfA, y="Notas", points="outliers", title="Grupo A")
fig.show()

fig = px.box(dfB, y="Notas", points="outliers", title="Grupo B")
fig.show()

Media del grupo A: 6.906666666666666
Media del grupo B: 6.593333333333334


El grupo A es más consistente dado a que el grupo B tiene notas más dispersas que se marcan como outliers

## Ejercicio 10. Mini proyecto de análisis

Contexto: dataset libre (notas, salarios, consumo, tiempos, etc.).

## Tareas

Cálculo de todas las medidas vistas.

Detección de outliers.

Al menos un diagrama de caja.

Informe final (10–15 líneas) respondiendo:

¿Qué está pasando en estos datos?

¿Qué conclusiones sacarías para tomar una decisión?

In [26]:
import pandas as pd
import numpy as np

### Dataset ejemplo

In [11]:
# Datos de ejemplo: horas de estudio semanales
datos = {
    "alumno": [
        "Ane", "Iker", "Jon", "Maialen", "Unai",
        "Nerea", "Aitor", "Leire", "Mikel", "Irati",
        "Asier", "June", "Oier", "Uxue", "Markel"
    ],
    "horas_estudio": [
        5, 6, 7, 6, 8,
        5, 6, 7, 6, 40,  # posible outlier
        5, 6, 7, 6, 8
    ]
}

df = pd.DataFrame(datos)
df


alumno  horas_estudio
0       Ane              5
1      Iker              6
2       Jon              7
3   Maialen              6
4      Unai              8
5     Nerea              5
6     Aitor              6
7     Leire              7
8     Mikel              6
9     Irati             40
10    Asier              5
11     June              6
12     Oier              7
13     Uxue              6
14   Markel              8

### Medidas de tendencia central

## TODO
- ¿Cuál representa mejor al alumno “típico”?
- ¿Por qué la media es mayor/menor que la mediana?


In [14]:
print("Media: ", df["horas_estudio"].mean())
print("Mediana: ", df["horas_estudio"].median())
print("Max: ", df["horas_estudio"].max())

Media:  8.533333333333333
Mediana:  6.0
Max:  40


La mediana representa mejor al "alumno típico" dado a que la máxima de 40h eleva mucho la media

### Medidas de dispersión

## TODO
- ¿El grupo es homogéneo o muy diverso?
- ¿Qué indica la desviación estándar?


In [18]:
print(df["horas_estudio"].value_counts())
print("Desviación estándard: ", df["horas_estudio"].std())

horas_estudio
6     6
5     3
7     3
8     2
40    1
Name: count, dtype: int64
Desviación estándard:  8.757581749275202


El grupo es homogéneo salvo por el outlier de 40h, la desviación estándard es alta y al estar por encima de la media indica que los valores están desplazados hacia la derecha

### Detección de outliers (método IQR)

### TODO
- ¿Cuántos outliers hay?
- ¿Qué explicación real podría haber detrás?


### Comparación con y sin outliers

In [21]:
Q1 = df["horas_estudio"].quantile(0.25)
Q3 = df["horas_estudio"].quantile(0.75)
IQR = Q3 - Q1

lim_inf = Q1 - 1.5 * IQR
lim_sup = Q3 + 1.5 * IQR

outliers = df[(df["horas_estudio"] < lim_inf) | (df["horas_estudio"] > lim_sup)]

print(outliers)

df_limpio = df[df["horas_estudio"] != 40]

print("Media: ", df_limpio["horas_estudio"].mean())
print("Mediana: ", df_limpio["horas_estudio"].median())
print("Max: ", df_limpio["horas_estudio"].max())
print("Desviación estándard: ", df_limpio["horas_estudio"].std())

  alumno  horas_estudio
9  Irati             40
Media:  6.285714285714286
Mediana:  6.0
Max:  8
Desviación estándard:  0.9944903161976937


Solo hay una outlier que o puede que haya metido mal el dato o puede que sea una excepción, sin ella los valores son muy homogeneos (std ~= 1) con ligera desviación a la izquierda

### TODO
- ¿Cómo cambia la percepción del grupo?
- ¿Qué medida usarías para informar al profesorado?


Sacar el outlier de las medias baja la media considerablemente, notificaría la información sin Irati y mencionaría la excepción de ella. De la estadística original usaría la mediana para dar una cifra más realista del comportamiento promedio

### Diagrama de caja

### TODO
- Describe el gráfico sin usar números
- ¿Dónde se concentra la mayoría?

In [23]:
fig = px.box(df, y="horas_estudio", points="outliers", title="Horas de estudio (con Irati)")
fig.show()

fig = px.box(df_limpio, y="horas_estudio", points="outliers", title="Horas de estudio (sin Irati)")
fig.show()

### STORYTELLING

Conclusión:

La mayoría del alumnado dedica alrededor de 6 horas semanales al estudio,
como indica la mediana.
La media es 8.53 horas debido al registro de 40 horas de Irati.

La dispersión de los datos es mayoritariamente homogenea,
lo que indica que el grupo es constante.

Se detectan valores atípicos que representan outliers,
los cuales elevan la interpretación general.



